File: Sources/LoeBalance/Settings/LoginViewModel.swift
import Foundation
import SwiftUI

@MainActor
final class LoginViewModel: ObservableObject {
    @Published var email = ""
    @Published var password = ""
    @Published private(set) var isSubmitting = false
    @Published var errorMessage: String?

    private let auth: AuthManager
    private let onSuccess: @MainActor () -> Void

    init(auth: AuthManager, onSuccess: @escaping @MainActor () -> Void = {}) {
        self.auth = auth
        self.onSuccess = onSuccess
    }

    func submit() async {
        guard !isSubmitting else { return }
        errorMessage = nil
        guard Self.isValidEmail(email) else {
            errorMessage = "Enter a valid email address."
            return
        }
        guard password.count >= 6 else {
            errorMessage = "Password must be at least 6 characters."
            return
        }

        isSubmitting = true
        var passwordCopy = password
        defer {
            passwordCopy.removeAll(keepingCapacity: false)
            password = ""
            isSubmitting = false
        }

        do {
            try await auth.login(email: email.trimmingCharacters(in: .whitespacesAndNewlines), password: passwordCopy)
            onSuccess()
        } catch {
            errorMessage = Self.message(for: error)
        }
    }

    private static func isValidEmail(_ value: String) -> Bool {
        let email = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let at = email.firstIndex(of: "@"), at > email.startIndex else { return false }
        let domain = email[email.index(after: at)...]
        return domain.contains(".") && !domain.hasPrefix(".") && !domain.hasSuffix(".")
    }

    private static func message(for error: Error) -> String {
        guard let appError = error as? AppError else {
            return "Unable to sign in."
        }
        return switch appError {
        case .loginRequired:
            "Sign-in is required."
        case .unauthorized:
            "Email or password is incorrect."
        default:
            "Unable to sign in."
        }
    }
}

File: windows/src/LoeBalance.Desktop.Wpf/ViewModels/LoginViewModel.cs
using LoeBalance.Core.Auth;
using LoeBalance.Core.Models;

namespace LoeBalance.Desktop.Wpf.ViewModels;

/// <summary>
/// Mirrors the macOS login view model: the same validation, the same messages, and the
/// password is cleared from memory as soon as the attempt finishes.
/// </summary>
public sealed class LoginViewModel
{
    private readonly AuthManager _auth;

    public LoginViewModel(AuthManager auth, Action onSuccess)
    {
        _auth = auth;
        OnSuccess = onSuccess;
    }

    public string Email { get; set; } = string.Empty;
    public string ErrorMessage { get; private set; } = string.Empty;
    public bool IsSubmitting { get; private set; }

    public event EventHandler? StateChanged;

    internal Action OnSuccess { get; }

    public async Task SubmitAsync(string password)
    {
        if (IsSubmitting) return;

        var email = Email.Trim();
        ErrorMessage = string.Empty;
        if (!IsValidEmail(email))
        {
            ErrorMessage = "Enter a valid email address.";
            RaiseStateChanged();
            return;
        }
        if (password.Length < 6)
        {
            ErrorMessage = "Password must be at least 6 characters.";
            RaiseStateChanged();
            return;
        }

        IsSubmitting = true;
        RaiseStateChanged();
        try
        {
            await _auth.LoginAsync(email, password);
            OnSuccess();
        }
        catch (AppException exception)
        {
            ErrorMessage = Message(exception);
        }
        catch (Exception)
        {
            ErrorMessage = "Unable to sign in.";
        }
        finally
        {
            IsSubmitting = false;
            // The caller clears the password box; the view model never keeps a copy.
            RaiseStateChanged();
        }
    }

    internal static bool IsValidEmail(string value)
    {
        var email = value.Trim();
        var at = email.IndexOf('@');
        if (at <= 0) return false;
        var domain = email[(at + 1)..];
        return domain.Contains('.') && !domain.StartsWith('.') && !domain.EndsWith('.');
    }

    private static string Message(AppException exception) => exception.Kind switch
    {
        AppErrorKind.Unauthorized => "Email or password is incorrect.",
        AppErrorKind.ApiEnvelope => "Sign-in is required.",
        AppErrorKind.Transport => "Unable to reach the server.",
        _ => "Unable to sign in."
    };

    private void RaiseStateChanged() => StateChanged?.Invoke(this, EventArgs.Empty);
}
